src/ClientIp.php

```diff
@@ -107,33 +107,78 @@ public function process(ServerRequestInterface $request, DelegateInterface $dele
      */
     private function getIp(ServerRequestInterface $request)
     {
-        if ($this->remote) {
-            $ip = file_get_contents('http://ipecho.net/plain');
+        $remoteIp = $this->getRemoteIp();
 
-            if (self::isValid($ip)) {
-                return $ip;
-            }
+        if (!empty($remoteIp)) {
+            // Found IP address via remote service.
+            return $remoteIp;
         }
 
-        $ip = null;
-        $server = $request->getServerParams();
+        $localIp = $this->getLocalIp($request);
 
-        if (!empty($server['REMOTE_ADDR']) && self::isValid($server['REMOTE_ADDR'])) {
-            $ip = $server['REMOTE_ADDR'];
+        if ($this->proxyIps && !in_array($localIp, $this->proxyIps)) {
+            // Local IP address does not point at a known proxy, do not attempt
+            // to read proxied IP address.
+            return $localIp;
         }
 
-        if (empty($this->proxyHeaders) || (!empty($this->proxyIps) && !in_array($ip, $this->proxyIps))) {
-            return $ip;
+        $proxiedIp = $this->getProxiedIp($request);
+
+        if (!empty($proxiedIp)) {
+            // Found IP address via proxy-defined headers.
+            return $proxiedIp;
         }
 
-        foreach ($this->proxyHeaders as $name) {
-            if ($request->hasHeader($name) && ($ip = self::getHeaderIp($request->getHeaderLine($name))) !== null) {
+        return $localIp;
```

Comment:
@oscarotero okay, functionality should be a match with existing code. Verify for me?